### PR TITLE
Add private subroutine fetch_by_seq_region_synonym

### PR DESCRIPTION
**Pull request by nwillhoft**
## Description 

There is code in SliceAdaptor::fetch_by_region() that has been marked for refactoring, specifically the code dealing with synonym/wildcard/fuzzy-matching. During work on creating a chromosome alias feature, I worked on refactoring this code into two new subroutines. This PR has simply added a new subroutine for code relating to fetching sequence regions by synonyms (and synonym-based wildcards). There are no calls to this subroutine or changes to fetch_by_region to make this first PR as simple as possible.

## Use case

This refactored code will be used by any SliceAdaptor::fetch_by_region request where the seq_region_name user argument is not an exact match to a seq_region_name attribute in the relevant core database.

I understand that this code is used extensively by Ensembl users and therefore a note here to say that these changes will of course be extensively tested. (Use cases to be added here during this work.)

## Benefits

In brief, the removal of superfluous code and making the code more readable.

During work on the chromosome alias feature, it became clear that the synonym/wildcard/fuzzy-match code needed to be refactored into subroutines to avoid large chunks of duplicate code being used in SliceAdaptor::fetch_by_region.

## Possible Drawbacks

None I can think of.

## Testing

_Have you added/modified unit tests to test the changes?_

Yes, tests have been added to `sliceAdaptor.t` to check the returned Slice object and to check an error is thrown when no synonym/wildcard match is found and fuzzy match has been set to false.

_If so, do the tests pass/fail?_

`sliceAdaptor.t` tests pass.

### CODE REVIEW

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm [hunk 1]
```diff
@@ -2728,4 +2728,74 @@ sub _build_circular_slice_cache {
...
+    $coord_system_name = $cs->name;
+    $syn_sql .= "AND cs.name = '" . $coord_system_name . "' ";
+  }
+  if (defined $version) {
+    $syn_sql .= "AND cs.version = '" . $version . "' ";
+  }
+  my $syn_sql_sth = $self->prepare($syn_sql);
+  $syn_sql_sth->bind_param(1, $seq_region_name, SQL_VARCHAR);
+  $syn_sql_sth->bind_param(2, $self->species_id(), SQL_INTEGER);
+  # print "SQL to run: $syn_sql\n";
```
Comment (magaliruffier) [inline modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm:2759]: These two lines can be removed

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm [hunk 2]
```diff
@@ -2728,4 +2728,74 @@ sub _build_circular_slice_cache {
...
+  # exit;
+  $syn_sql_sth->execute();
+  my ($new_name, $new_coord_system, $new_version);
+  $syn_sql_sth->bind_columns( \$new_name, \$new_coord_system, \$new_version);
+  # print Dumper( $syn_sql_sth );
```
Comment (magaliruffier) [inline modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm:2764]: remove

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm [hunk 3]
```diff
@@ -2728,4 +2728,74 @@ sub _build_circular_slice_cache {
...
+  if($syn_sql_sth->fetch){
+    if ((not defined($cs)) || ($cs->name eq $new_coord_system && $cs->version eq $new_version)) {
+        # print "New values: $new_name, $new_coord_system, $new_version\n";
```
Comment (magaliruffier) [inline modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm:2767]: remove

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm [hunk 4]
```diff
@@ -2728,4 +2728,74 @@ sub _build_circular_slice_cache {
...
+        print "New values: $new_coord_system, $new_name, $start, $end, $strand, $new_version, $no_fuzz\n";
```
Comment (magaliruffier) [inline modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm:2768]: do we need this here?

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm [hunk 5]
```diff
@@ -2728,4 +2728,74 @@ sub _build_circular_slice_cache {
...
+        return $self->fetch_by_region($new_coord_system, $new_name, $start, $end, $strand, $new_version, $no_fuzz);
+    }
+  } else {
+    # Try wildcard searching if no exact synonym was found
+    print "DEBUG: No exact synonym found for $seq_region_name - try wildcard searching...\n";
```
Comment (magaliruffier) [inline modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm:2773]: remove

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm [hunk 6]
```diff
@@ -2728,4 +2728,74 @@ sub _build_circular_slice_cache {
...
+    my $escape_char = $self->dbc->db_handle->get_info(14);
+    $escaped_seq_region_name =~ s/([_%])/$escape_char$1/g;
+    $syn_sql_sth->bind_param(1, "$escaped_seq_region_name%", SQL_VARCHAR);
+    $syn_sql_sth->bind_param(2, $self->species_id(), SQL_INTEGER); 
+    $syn_sql_sth->execute();
+    $syn_sql_sth->bind_columns( \$new_name, \$new_coord_system, \$new_version);
+
+    if($syn_sql_sth->fetch){
+      if ((not defined($cs)) || ($cs->name eq $new_coord_system && $cs->version eq $new_version)) {
+          print "Returning: $self->fetch_by_region($new_coord_system, $new_name, $start, $end, $strand, $new_version, $no_fuzz)\n";
```
Comment (magaliruffier) [inline modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm:2785]: remove

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm [hunk 7]
```diff
@@ -2728,4 +2728,74 @@ sub _build_circular_slice_cache {
...
+          return $self->fetch_by_region($new_coord_system, $new_name, $start, $end, $strand, $new_version, $no_fuzz);
+      } elsif ($cs->name ne $new_coord_system) {
+          warning("Searched for a known feature on coordinate system: ".$cs->dbID." but found it on: ".$new_coord_system.
+          "\n No result returned, consider searching without coordinate system or use toplevel.");
+          return;
+      }
+      
+    } else {
+      warning("DEBUG: No result from SQL query. No synonym match found for $seq_region_name\n");
```
Comment (magaliruffier) [inline modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm:2794]: remove

**Review by magaliruffier — CHANGES_REQUESTED** [review]
The overall method does what it says

Two main changes:
- reduce the number of prints
Think about the number of times this method might be called and whether it is useful for the user to see all that information every time
- add a test case for this method on its own
If we ever want to update the fetch_by_region method, we can isolate components as long as they are tested independently

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm (hunk 1 above)
Comment (nwillhoft) [reply modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm]: Removed.

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm (hunk 2 above)
Comment (nwillhoft) [reply modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm]: Removed.

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm (hunk 3 above)
Comment (nwillhoft) [reply modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm]: Removed.

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm (hunk 4 above)
Comment (nwillhoft) [reply modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm]: Not anymore, removed.

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm (hunk 5 above)
Comment (nwillhoft) [reply modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm:2768]: Removed.

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm (hunk 6 above)
Comment (nwillhoft) [reply modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm]: Removed.

File: modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm (hunk 7 above)
Comment (nwillhoft) [reply modules/Bio/EnsEMBL/DBSQL/SliceAdaptor.pm]: Removed.

**Comment by nwillhoft** [comment]
Have removed extra print statements used for previous debugging. Will now work on adding test case specific to this method.

**Comment by nwillhoft** [comment]
Added a couple of test cases for this new method to `sliceAdaptor.t`.

**Review by magaliruffier — APPROVED** [review]
Previous review comments have been addressed and the test cases are comprehensive

File: modules/t/sliceAdaptor.t [hunk 8]
```diff
@@ -56,6 +55,29 @@ ok($slice->seq_region_length == 62842997);
...
+  my $synonym = "anoth_20";
+  my $start = "1";
+  my $end = "1e6";
+  my $strand = "1";
+  my $target_seq_name = "20";
+  my $target_cs_name = "chromosome";
+  my $synonym_slice = $slice_adaptor->_fetch_by_seq_region_synonym( undef, $synonym, $start, $end, $strand, undef, undef );
+  test_slice("${target_seq_name}:${start}-${end}", $synonym_slice, $target_cs_name, $target_seq_name, $start, $end, $strand);
+}
+
```
Comment (ens-bwalts) [inline modules/t/sliceAdaptor.t:72]: I would add one more test here to get through the synonym-wildcard code path (maybe make one for `$synonym = "anoth"`

Comment (ens-bwalts) [reply modules/t/sliceAdaptor.t:72]: Otherwise, this looks good

Comment (nwillhoft) [reply modules/t/sliceAdaptor.t:72]: Thanks Brandon, will add the suggested synonym-wildcard test in.

**Comment by mira13** [comment]
We are not checking any input arguments here. If seq_region_name or start or end doesn't exist - it will be crash or unexpected behaviour without any readable error msg.

**Comment by nwillhoft** [comment]
> We are not checking any input arguments here. If seq_region_name or start or end doesn't exist - it will be crash or unexpected behaviour without any readable error msg.

Thanks I'll add in checks for input arguments.

**Review by ens-bwalts — APPROVED** [review]
Test coverage is improved.

**Pull request merged**